src/passes/TypeGeneralizing.cpp

```diff
@@ -323,81 +394,469 @@ struct TransferFn : OverriddenVisitor<TransferFn> {
+    // that propagate the tested value, we need to propagate the existing
+    // requirement instead. Like br_if, these instructions will require extra
+    // basic blocks on the branches that drop values.
+    WASM_UNREACHABLE("TODO");
+  }
+
+  void visitStructNew(StructNew* curr) {
+    // We cannot yet generalize allocations. Push requirements for the types
+    // needed to initialize the struct.
+    pop();
+    if (!curr->isWithDefault()) {
+      auto type = curr->type.getHeapType();
+      for (const auto& field : type.getStruct().fields) {
+        if (field.type.isRef()) {
+          push(field.type);
+        }
+      }
+    }
+  }
+
+  HeapType
+  generalizeStructType(HeapType type,
+                       Index index,
+                       std::optional<Type> reqFieldType = std::nullopt) {
+    // Find the most general struct type for which this access could be valid,
+    // i.e. the most general supertype that still has a field at the given index
+    // where the field is a subtype of the required type, if any.
+    while (true) {
+      auto candidateType = type.getDeclaredSuperType();
+      if (!candidateType) {
+        // Cannot get any more general.
+        break;
+      }
+      const auto& candidateFields = candidateType->getStruct().fields;
+      if (candidateFields.size() <= index) {
+        // Cannot get any more general and still have a field at the necessary
+        // index.
+        break;
+      }
+      if (reqFieldType) {
+        auto candidateFieldType = candidateFields[index].type;
+        if (candidateFieldType != *reqFieldType &&
+            Type::isSubType(*reqFieldType, candidateFieldType)) {
+          // Cannot generalize without violating the requirements on the field.
+          break;
+        }
+      }
+      type = *candidateType;
+    }
+    return type;
+  }
+
+  void visitStructGet(StructGet* curr) {
+    auto type = curr->ref->type.getHeapType();
+    if (type.isBottom()) {
+      // This will be emitted as unreachable. Do not require anything of the
+      // input, except that the ref remain bottom.
+      clearStack();
+      push(Type(HeapType::none, Nullable));
+      return;
+    }
+    std::optional<Type> reqFieldType;
+    if (curr->type.isRef()) {
+      reqFieldType = pop();
+    }
+    auto generalized = generalizeStructType(type, curr->index, reqFieldType);
+    push(Type(generalized, Nullable));
+  }
+
+  void visitStructSet(StructSet* curr) {
+    auto type = curr->ref->type.getHeapType();
+    if (type.isBottom()) {
+      // This will be emitted as unreachable. Do not require anything of the
+      // input except that the ref remain bottom.
+      clearStack();
+      push(Type(HeapType::none, Nullable));
+      if (curr->value->type.isRef()) {
+        push(Type::none);
+      }
+      return;
+    }
+    auto generalized = generalizeStructType(type, curr->index);
+    push(Type(generalized, Nullable));
+    push(generalized.getStruct().fields[curr->index].type);
```

Comment:
Nothing is popped here at all - how does this work? This seems to only read from static locations (that are not modified during the analysis).

I would expect to see that changes to the reference affect the value, and changes to the value affect the reference. Can that not work, or is there a TODO here?

Reply:
Right, `struct.set` has no return value, so there is nothing to pop. Since mutable struct fields are invariant, it's always valid to generalize the type of the struct we are setting as long as it still has a field at the correct index; that field will have the same type in the generalized struct type as in the original struct type. We can also generalize the value being set up to the type of the field being assigned to.

Reply:
I see, thanks. What I was missing is that this is a write, so any supertype with the field is ok to use, and so there aren't really any "dynamic" constraints here. That is, the more the ref is generalized does not help the value generalize, or vice versa.